### PR TITLE
refactor: update renderQuery to Suspense + Relay Hooks #5297

### PR DESCRIPTION
**Pull request by JimmyLv**
Create the PR early to make sure my first refactoring case is on the correct road, welcome code review earlier!

There are 27 remaining places that use `<QueryRender`, and these are my TODO list, which I can use in the meantime to familiarize myself with Parabol's user functionalities.

- [x] [self-check Code Review list](https://github.com/ParabolInc/parabol/blob/master/docs/codeReview.md)
- [x] ~~add basic Playwright functional tests to reduce manual test when refactoring~~ currently, there are no authenticated test cases, which need more effort (out of scope) via [packages/integration-tests/README.md](https://github.com/ParabolInc/parabol/blob/master/packages/integration-tests/README.md)
- [ ] ~~should refactoring the `createFragmentContainer` into `useFragment` in this PR? via [Realy Hooks](https://relay.dev/docs/migration-and-compatibility/relay-hooks-and-legacy-container-apis/#fragment-container--usefragment)~~
   - [ ] ~~also includes `createPaginationContainer` and `createRefetchContainer`~~
- [x] extract some common pattern of components, e.g. `<DefaultSuspense />` with fallback `Loader` as props, and `useDefaultPreloadedQuery` -> `renderLoader(options)`
- [x] check all loading spinner and related data-loading user experiences

functional use cases:

- [x] InvitationRoot.tsx
- [x] InvoiceRoot.tsx
- [x] JiraScopingSearchFilterMenuRoot.tsx
- [x] JiraScopingSearchResultsRoot.tsx
- [x] MassInvitationTokenLinkRoot.tsx
- [x] MeetingCardOptionsMenuRoot.tsx
- [x] MeetingSummaryEmailRootSSR.tsx
- [x] NewMeetingRoot.tsx
- [x] NewMeetingSummaryRoot.tsx
- [x] OrganizationRoot.tsx
- [x] OrganizationsRoot.tsx
- [x] OrgBillingRoot.tsx
- [x] OrgMembersRoot.tsx
- [x] ParabolScopingSearchResultsRoot.tsx
- [x] PokerTemplateListOrgRoot.tsx
- [x] PokerTemplateListPublicRoot.tsx
- [x] ReflectTemplateListOrgRoot.tsx
- [x] ReflectTemplateListPublicRoot.tsx
- [x] SuggestMentionableUsersRoot.tsx
- [x] TaskFooterIntegrateMenuRoot.tsx
- [x] TaskFooterTeamAssigneeMenuRoot.tsx
- [x] TaskFooterUserAssigneeMenuRoot.tsx
- [x] TeamIntegrationsRoot.tsx
- [x] TeamInvitationRoot.tsx
- [x] TeamSettingsRoot.tsx
- [x] UnpaidTeamModalRoot.tsx
- [x] UserProfileRoot.tsx
- [x] ViewerNotOnTeamRoot.tsx

### CODE REVIEW

File: packages/client/components/InvitationLink.tsx
Comment (Dschoordsch) [inline packages/client/components/InvitationLink.tsx:16]: -1 The query should be renamed to the place where it lives, so in this case
```
query InvitationLinkQuery($token: ID!) {
```

File: packages/client/components/InvitationLink.tsx
Comment (Dschoordsch) [inline packages/client/components/InvitationLink.tsx:23]: Fun fact, I googled what that means, this was my first result: https://www.parabol.co/blog/upgrade-react-to-graphql-relay-hooks/ Kudos @mattkrick 
TL;DR: we need this until we get rid of `createFragmentContainer`

File: packages/client/components/TeamInvitation.tsx
Comment (Dschoordsch) [inline packages/client/components/TeamInvitation.tsx:17]: -1 see above, please rename, same in the files following

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
Looks good so far!

File: packages/client/components/TeamInvitation.tsx
Comment (JimmyLv) [reply packages/client/components/TeamInvitation.tsx:17]: sure, thanks Georg @Dschoordsch

File: packages/client/components/InvitationLink.tsx
Comment (JimmyLv) [reply packages/client/components/InvitationLink.tsx:16]: Since it's not on the Root page that needs to be modified, this should just be a naming convention issue, right? Is there a similar specification document like this?

**Comment by JimmyLv** [comment]
@Dschoordsch Hey Georg! I have refactored all the component files related to `QueryRenderer` and removed the`renderQuery`, could you please help to review again? 🤩
I have checked the availability of the functionalities _before and after_ (when doing each Git commit), but can't guarantee that all the refactoring is exactly as expected by the product.

This is my first PR and any feedback/suggestions are welcome, thank you so much! ✌️

File: packages/client/components/NewMeeting.tsx
Comment (JimmyLv) [inline packages/client/components/NewMeeting.tsx:150]: actually, I'm not sure if the `retry` is still needed here, since the `refetch` is already set inside the `useQueryLoaderNow`

File: packages/client/modules/teamDashboard/containers/TeamIntegrationsRoot/TeamIntegrationsRoot.tsx
Comment (JimmyLv) [inline packages/client/modules/teamDashboard/containers/TeamIntegrationsRoot/TeamIntegrationsRoot.tsx:16]: this `retry` also need some attention 🤔

File: packages/client/modules/userDashboard/containers/OrgMembers/OrgMembersRoot.tsx
Comment (JimmyLv) [inline packages/client/modules/userDashboard/containers/OrgMembers/OrgMembersRoot.tsx:29]: for these Suspense `fallback` here, not quite sure if I need to add back the `Loader.Size` Loader component at this level

File: packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx
Comment (JimmyLv) [inline packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx:40]: This pattern would be simpler without involving modifications to the child component, e.g. `OrgBilling` (especially if it involves the `createPaginationContainer` and `createRefetchContainer` HOC components)

Because it uses `usePreloadedQuery` to request data, I call it the `Container` of component (compare to `Wrapper`)

File: packages/client/components/JiraScopingSearchResultsRoot.tsx
Comment (JimmyLv) [inline packages/client/components/JiraScopingSearchResultsRoot.tsx:31]: Not quite sure if I need to pass the `eventId` in from the `fallback()` parameter of the `ErrorBoundary`

File: packages/client/components/JiraScopingSearchResults.tsx
Comment (JimmyLv) [inline packages/client/components/JiraScopingSearchResults.tsx:89]: `<MockScopingList />` here is a duplicate with the `fallback` of Suspense at the upper level

File: packages/client/modules/outcomeCard/components/OutcomeCardAssignMenu/TaskFooterTeamAssigneeMenu.tsx
Comment (Dschoordsch) [inline packages/client/modules/outcomeCard/components/OutcomeCardAssignMenu/TaskFooterTeamAssigneeMenu.tsx:196]: -1 please convert it fully away from `createFragmentContainer`

File: packages/client/modules/outcomeCard/components/TaskFooterTeamAssigneeMenuRoot.tsx
Comment (Dschoordsch) [inline packages/client/modules/outcomeCard/components/TaskFooterTeamAssigneeMenuRoot.tsx:24]: -1 `renderQuery` had a loading component, probably should be `MockList`

File: packages/client/modules/outcomeCard/components/TaskFooterUserAssigneeMenuRoot.tsx
Comment (Dschoordsch) [inline packages/client/modules/outcomeCard/components/TaskFooterUserAssigneeMenuRoot.tsx:31]: -1 `renderQuery` had a fallback

File: packages/client/modules/summary/components/NewMeetingSummaryRoot.tsx
Comment (Dschoordsch) [inline packages/client/modules/summary/components/NewMeetingSummaryRoot.tsx:17]: -1 `renderQuery` had a fallback

File: packages/client/modules/teamDashboard/components/TeamSettingsRoot.tsx
Comment (Dschoordsch) [inline packages/client/modules/teamDashboard/components/TeamSettingsRoot.tsx:14]: -1 `renderQuery` had a fallback

File: packages/client/modules/teamDashboard/containers/TeamIntegrationsRoot/TeamIntegrationsRoot.tsx
Comment (Dschoordsch) [reply packages/client/modules/teamDashboard/containers/TeamIntegrationsRoot/TeamIntegrationsRoot.tsx:16]: +1 should work for now. A nicer solution would be to have a version of `useQueryLoaderNow` which exposes `retry()`
```
export const useQueryLoaderNowWithRetry = <TQuery extends OperationType>(
   ...
   return {queryRef, retry: refreshQuery}
}
const useQueryLoaderNow = ...
  const {queryRef} = useQueryLoaderNowWithRetry(...)
  return queryRef
}
```

File: packages/client/modules/teamDashboard/containers/UnpaidTeamModal/UnpaidTeamModalRoot.tsx
Comment (Dschoordsch) [inline packages/client/modules/teamDashboard/containers/UnpaidTeamModal/UnpaidTeamModalRoot.tsx:16]: -1 `renderQuery` had a fallback

File: packages/client/utils/relay/renderFallback.tsx
Comment (Dschoordsch) [inline packages/client/utils/relay/renderFallback.tsx:1]: -1 the file name is not ideal, `renderLoader` would be more fitting

File: packages/client/modules/userDashboard/containers/Organization/OrganizationRoot.tsx
Comment (Dschoordsch) [inline packages/client/modules/userDashboard/containers/Organization/OrganizationRoot.tsx:17]: -1 `renderQuery` had a fallback

File: packages/client/modules/userDashboard/containers/OrgMembers/OrgMembersRoot.tsx
Comment (Dschoordsch) [reply packages/client/modules/userDashboard/containers/OrgMembers/OrgMembersRoot.tsx:29]: -1 I think we should have a `MockList` fallback here

File: packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx
Comment (Dschoordsch) [inline packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx:49]: -1 If you touch this already, please convert the `createFragmentContainer` call to make it more readable overall

File: packages/client/modules/userDashboard/components/UserSettingsForm/UserSettingsForm.tsx
Comment (Dschoordsch) [inline packages/client/modules/userDashboard/components/UserSettingsForm/UserSettingsForm.tsx:1]: +1 It probably made sense to move the logic out, but it makes review harder if unrelated changes are intermingled. No need to revert this, but just a note for the future.

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
I'm not sure if you had a reason to remove the render fallback in many cases?

File: packages/client/components/AvatarInput.tsx
Comment (JimmyLv) [reply packages/client/components/AvatarInput.tsx:14]: since it is too trivial so I just take it easy to fix 😁

https://github.com/ParabolInc/parabol/issues/6169#issuecomment-1075471314

File: packages/client/components/JiraScopingSearchResultsRoot.tsx
Comment (JimmyLv) [reply packages/client/components/JiraScopingSearchResultsRoot.tsx:31]: cool, I just saw the `render()` method of `ErrorComponent`, so by default, we don't need the error fallback.

File: packages/client/components/JiraScopingSearchResults.tsx
Comment (JimmyLv) [reply packages/client/components/JiraScopingSearchResults.tsx:89]: okay, so I also remove the related comments here.

File: packages/client/components/ParabolScopingSearchResultsRoot.tsx
Comment (JimmyLv) [reply packages/client/components/ParabolScopingSearchResultsRoot.tsx:71]: `...Container` here, because I'm not going to deal with the `createPaginationContainer` inside the `ParabolScopingSearchResult` component yet, maybe we can consider refactoring it to `usePaginationFragment` in #6283

Comment (JimmyLv) [reply packages/client/components/ParabolScopingSearchResultsRoot.tsx:71]: but I did the refactoring to replace createFragmentContainer with `useFragment` for ParabolScopingSearchResultsRoot

File: packages/client/hooks/useQueryLoaderNow.tsx
Comment (JimmyLv) [reply packages/client/hooks/useQueryLoaderNow.tsx]: this one for MeetingSummaryEmailRootSSR https://github.com/ParabolInc/parabol/commit/8b474c9a45e9ca2b450a990b2416ee091f69e44e

File: packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx
Comment (JimmyLv) [reply packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx:48]: What I'm thinking of is removing the `Suspense`, but not quite sure if the environment specified here works on SSR.  @mattkrick have any ideas?

Another thing is: about the feature of sending emails, how do I go about verifying it in the PR phase?

**Comment by JimmyLv** [comment]
Hi @Dschoordsch I have solved most of the feedback, could you please help me to re-review it? 🙋‍♂️

File: packages/client/modules/userDashboard/components/UserSettingsForm/UserSettingsForm.tsx
Comment (Dschoordsch) [reply packages/client/modules/userDashboard/components/UserSettingsForm/UserSettingsForm.tsx:1]: If there are dependencies, one PR can target the feature branch of another PR instead of master. Once the first branch is merged, GitHub will change the target branch of the second one to master.

File: packages/client/components/JiraScopingSearchResultsRoot.tsx
Comment (Dschoordsch) [reply packages/client/components/JiraScopingSearchResultsRoot.tsx]: Yes, that's also something I learned while doing this review

File: packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx
Comment (Dschoordsch) [reply packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx]: There is a dummy mail provider which just prints the email's plain text to stdout. I guess it would be easy to modify it to print the html instead.

File: packages/client/modules/userDashboard/components/UserSettingsForm/UserSettingsForm.tsx
Comment (JimmyLv) [reply packages/client/modules/userDashboard/components/UserSettingsForm/UserSettingsForm.tsx:1]: cool, make sense! thanks @Dschoordsch

**Comment by JimmyLv** [comment]
@Dschoordsch Hey Georg, how about other feedback? and should I move this PR into the maintainer review column?

**Comment by Dschoordsch** [comment]
@JimmyLv sorry it takes so long. I got stuck with the review on the 2 places where you created a container as a double wrapper. I'm not happy getting it merged like that, but didn't had the time to look into a neater solution as I'm also lacking experience with the relay hooks, so I don't know how these interact with pagination. I will set aside some time today to check.

**Review by Dschoordsch — APPROVED** [review]
I think having suspense in the email rendering server side does not do any harm and at least is consistent with the rest of the components [packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx](https://github.com/ParabolInc/parabol/pull/6251/files/ac2c01a7c5b9c18949be374553e4d5a55846d1b7#diff-b83de9a3cc74b538222bad9310572bd10997c035ac8e10beb5aa178fc6d10fdd). 
I couldn't find a nicer solution for
* [packages/client/components/ParabolScopingSearchResultsRoot.tsx](https://github.com/ParabolInc/parabol/pull/6251/files/789d0c0f7cfaff2541d642ff8bd82a1065a8f7bd#diff-73ba6307f98e0ca59f8244045cacc31b699818b7ff2ce086b5f4712983f287f3)
* [packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx](https://github.com/ParabolInc/parabol/pull/6251/files/ac2c01a7c5b9c18949be374553e4d5a55846d1b7#diff-19053e2335214fa597bca5c90280d4f99f71cc21cde3d98f8bb81b70b4cb283f)

without refactoring to `usePaginationFragment` which requires even more refactoring.

Let's see what @mattkrick has to say about it.

**Comment by JimmyLv** [comment]
@Dschoordsch Thanks for the detailed review! 

about the `usePaginationFragment`, I think it would be better if we consider refactoring these Legacy Container APIs to Relay Hooks together (maybe in #6283). I found several areas that needed to be refactored, including `createPaginationContainer` and `createRefetchContainer`.

ref: [Relay Hooks and Legacy Container APIs | Relay](https://relay.dev/docs/migration-and-compatibility/relay-hooks-and-legacy-container-apis/#refetch-container--userefetchablefragment)

File: packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx
Comment (mattkrick) [inline packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx:40]: -1 i believe we'll have to change this because now we're importing more than just a type, we're importing an object.

This is problematic because to build `meetingSummaryEmailRootSsrQuery`, relay-compiler needs a graphql schema & to build a graphql schema, graphql must traverse the /graphql/public dir, which eventually depends on this file. This creates a circular dependency.

I'd suggest using a lazy-load pattern. See https://github.com/ParabolInc/parabol/blob/0cb19afc1043501a55a990e3a838c1792840e1bb/packages/client/components/DemoMeetingRoot.tsx#L30

File: packages/client/modules/meeting/components/PokerTemplateListOrgRoot.tsx
Comment (mattkrick) [inline packages/client/modules/meeting/components/PokerTemplateListOrgRoot.tsx:16]: -1 we don't want to trigger a network request until `isActive` is true. easiest fix is to probably move the `isActive` logic to the parent.

TEST
- [ ] open dev tools -> network -> websocket & monitor traffic
- [ ] open the poker meeting settings
- [ ] this query should _not_ get fetched until the organization tab is clicked

File: packages/client/modules/meeting/components/PokerTemplateListPublicRoot.tsx
Comment (mattkrick) [inline packages/client/modules/meeting/components/PokerTemplateListPublicRoot.tsx:16]: -1 same as above, network request only on isActive

File: packages/client/modules/meeting/components/ReflectTemplateListOrgRoot.tsx
Comment (mattkrick) [inline packages/client/modules/meeting/components/ReflectTemplateListOrgRoot.tsx:16]: -1 same as above, network request only on isActive

File: packages/client/modules/meeting/components/ReflectTemplateListPublicRoot.tsx
Comment (mattkrick) [inline packages/client/modules/meeting/components/ReflectTemplateListPublicRoot.tsx:16]: -1 same as above, network request only on isActive

File: packages/client/modules/outcomeCard/components/TaskFooterTeamAssigneeMenuRoot.tsx
Comment (mattkrick) [inline packages/client/modules/outcomeCard/components/TaskFooterTeamAssigneeMenuRoot.tsx:25]: +2 nice touch adding a fallback here!

File: packages/client/modules/summary/components/NewMeetingSummaryRoot.tsx
Comment (mattkrick) [inline packages/client/modules/summary/components/NewMeetingSummaryRoot.tsx:25]: +1 `withRouter` is deprecated, you could replace it with `useRouter` if you like, although if this PR is too big, I understand

File: packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx
Comment (mattkrick) [reply packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx:52]: makes sense, I'll backlog that work in https://github.com/ParabolInc/parabol/issues/6431

**Review by mattkrick — CHANGES_REQUESTED** [review]
nice work! couple small things, let me know if they give you trouble fixing

File: packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx
Comment (JimmyLv) [reply packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx]: make sense! I updated to use `useLazyLoadQuery` to load data directly and avoid circular dependency

File: packages/client/modules/meeting/components/PokerTemplateListOrgRoot.tsx
Comment (JimmyLv) [reply packages/client/modules/meeting/components/PokerTemplateListOrgRoot.tsx:16]: good point! This is the difference between component-based usage and React hook, and I've updated it!

**Comment by JimmyLv** [comment]
@mattkrick Thanks for the review! Victory is at hand 🤩
I did update with feedback, including `useLazyLoadQuery` and network requests for Poker and Reflect TemplateList, etc. 
Then I did a merge with the master and I think `git precommit` hook did some automatic formatting while handling the conflict, hopefully it will not be too disturbing.

**Comment by mattkrick** [comment]
LGTM! no worries about the formatting, we'll run https://github.com/ParabolInc/parabol/issues/6314 in the next sprint & get the whole repo formatted correctly so we don't encounter those things anymore

**Pull request merged**
